crypto/key.go

```diff
@@ -0,0 +1,205 @@
+/*
+	This file is part of go-ethereum
+
+	go-ethereum is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Lesser General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	go-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU Lesser General Public License
+	along with go-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @authors
+ * 	Gustav Simonsson <gustav.simonsson@gmail.com>
+ * @date 2015
+ *
+ */
+
+package crypto
+
+import (
+	"bytes"
+	"code.google.com/p/go-uuid/uuid"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	crand "crypto/rand"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"runtime"
+	"strings"
+	"time"
+)
+
+type Key struct {
+	Id    *uuid.UUID // Version 4 "random" for unique id not derived from key data
+	Flags [4]byte    // RFU
+	// we only store privkey as pubkey/address can be derived from it
+	// privkey in this struct is always in plaintext
+	PrivateKey *ecdsa.PrivateKey
+}
+
+type PlainKeyJSON struct {
+	Id         string
+	Flags      string
+	PrivateKey string
+}
+
+type CipherJSON struct {
+	Salt       string
+	IV         string
+	CipherText string
+}
+
+type EncryptedKeyJSON struct {
+	Id     string
+	Flags  string
+	Crypto CipherJSON
+}
+
+func (k *Key) Address() []byte {
+	pubBytes := FromECDSAPub(&k.PrivateKey.PublicKey)
+	return Sha3(pubBytes)[12:]
+}
+
+func (k *Key) MarshalJSON() (j []byte, err error) {
+	stringStruct := PlainKeyJSON{
+		k.Id.String(),
+		hex.EncodeToString(k.Flags[:]),
+		hex.EncodeToString(FromECDSA(k.PrivateKey)),
+	}
+	j, err = json.Marshal(stringStruct)
+	return j, err
+}
+
+func (k *Key) UnmarshalJSON(j []byte) (err error) {
+	keyJSON := new(PlainKeyJSON)
+	err = json.Unmarshal(j, &keyJSON)
+	if err != nil {
+		return err
+	}
+
+	u := new(uuid.UUID)
+	*u = uuid.Parse(keyJSON.Id)
+	if *u == nil {
+		err = errors.New("UUID parsing failed")
+		return err
+	}
+	k.Id = u
+
+	flagsBytes, err := hex.DecodeString(keyJSON.Flags)
```

Comment:
True, right now there are no such requirements. However, with multiple ethereum implementations it would be nice to have one format so one can move keys around. That's why I use JSON and hex encoding and also a fixed size for the flag bytes.

I suppose a variable array and base64 encoding is no less easy to integrate in other implementations, so we could for sure use that.

One benefit of hex encoding compared to base64 would be for the flags array though, as it's easier to inspect the hex to see what is set. Currently it's not actually used though, so maybe we should actually remove it for now to make the struct simpler.